### PR TITLE
sys-devel/mlir: new subpackage from LLVM-13

### PR DESCRIPTION
**Pull request by dzejej**
This package is needed by LLVM based fortran compiler

Package-Manager: Portage-3.0.30, Repoman-3.0.3
Signed-off-by: Janusz Andrzejewski <janusz.andrzejewski@gmail.com>

### CODE REVIEW

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+DEPEND="
+	sys-devel/llvm:13=[${MULTILIB_USEDEP}]
+"
+RDEPEND="${DEPEND}"
+
+LLVM_MAX_SLOT=13
+
+src_unpack() {
+	unpack llvmorg-13.0.0.tar.gz
+	mv "${WORKDIR}/llvm-project-llvmorg-13.0.0/mlir" "${WORKDIR}/${P}"
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:27]: `|| die` on external commands which may fail

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		-DCMAKE_PREFIX_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/lib64/cmake/llvm/"
+		-DLLVM_LINK_LLVM_DYLIB=ON
+		-DLLVM_BUILD_LLVM_DYLIB=ON
+		-DMLIR_LINK_MLIR_DYLIB=ON
+		-DLLVM_BUILD_TOOLS=ON
+		-DLLVM_BUILD_UTILS=ON
+		-DLLVM_ENABLE_PIC=ON
+		-DMLIR_INCLUDE_TESTS=OFF
+		-Wno-dev
+		)
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:46]:
```suggestion
	)
```

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		-DBUILD_SHARED_LIBS:BOOL=ON
+		-DLLVM_LINK_LLVM_DYLIB:BOOL=ON
+		-DCMAKE_PREFIX_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/lib64/cmake/llvm/"
+		-DLLVM_LINK_LLVM_DYLIB=ON
+		-DLLVM_BUILD_LLVM_DYLIB=ON
+		-DMLIR_LINK_MLIR_DYLIB=ON
+		-DLLVM_BUILD_TOOLS=ON
+		-DLLVM_BUILD_UTILS=ON
+		-DLLVM_ENABLE_PIC=ON
+		-DMLIR_INCLUDE_TESTS=OFF
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:44]: Do tests run right now?

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+multilib_src_configure() {
+	local mycmakeargs=(
+		-GNinja
+		-DCMAKE_BUILD_TYPE=Release
+		-DCMAKE_INSTALL_PREFIX="${EPREFIX}/usr/lib/llvm/${SLOT}"
+		-DBUILD_SHARED_LIBS:BOOL=ON
+		-DLLVM_LINK_LLVM_DYLIB:BOOL=ON
+		-DCMAKE_PREFIX_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/lib64/cmake/llvm/"
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:37]:
```suggestion
		-DCMAKE_PREFIX_PATH="${EPREFIX}/usr/lib/llvm/${SLOT}/$(get_libdir)/cmake/llvm/"
```

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+EAPI=7
+
+CMAKE_ECLASS=cmake
+
+inherit cmake-multilib llvm
+
+DESCRIPTION="Multi-Level IR Compiler Framework for LLVM"
+HOMEPAGE="https://mlir.llvm.org/"
+SRC_URI=" https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-13.0.0.src.tar.gz -> llvmorg-13.0.0.tar.gz"
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:12]: You're hard coding version numbers here:
```suggestion
SRC_URI=" https://github.com/llvm/llvm-project/releases/download/llvmorg-${PV}/llvm-${PV}.src.tar.gz"
```

Why rename to 'llvmorg'? Actually, isn't this what `llvm.org.eclass` is for? (see the other LLVM ebuilds)

File: sys-devel/mlir/mlir-13.0.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+DEPEND="
+	sys-devel/llvm:13=[${MULTILIB_USEDEP}]
+"
+RDEPEND="${DEPEND}"
+
+LLVM_MAX_SLOT=13
+
+src_unpack() {
+	unpack llvmorg-13.0.0.tar.gz
```
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:26]:
```suggestion
	unpack llvm-${PV}.src.tar.gz
```

File: sys-devel/mlir/mlir-13.0.0.ebuild (hunk 5 above)
Comment (thesamesam) [inline sys-devel/mlir/mlir-13.0.0.ebuild:12]:
```suggestion
SRC_URI="https://github.com/llvm/llvm-project/releases/download/llvmorg-13.0.0/llvm-13.0.0.src.tar.gz -> llvmorg-13.0.0.tar.gz"
```

**Review by thesamesam — COMMENTED** [review]
Left some provisional comments but needs mgorny here.

Thanks for looking into this and Flang.

**Comment by thesamesam** [comment]
Oh, and it would be okay to do this and Flang in the same PR as you're only adding it for Flang anyway.

**Comment by stefson** [comment]
it's somewhat common sense to have a full llvm-13 toolchain available before emerging mlir on top of it, but still the ebuild should pull in the parts of the toolchain it needs. Got this error when only llvm-12 toolchain is available: 

```
-- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR
-- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR - Failed
CMake Error at cmake/modules/AddMLIR.cmake:149 (get_target_export_arg):
  Unknown CMake command "get_target_export_arg".
Call Stack (most recent call first):
  cmake/modules/AddMLIR.cmake:135 (add_mlir_library_install)
  lib/Analysis/CMakeLists.txt:21 (add_mlir_library)


-- Configuring incomplete, errors occurred!
```

on another system with llvm-13 available, there were no problems. So please pull in the correct part of the toolchain.

**Comment by dzejej** [comment]
The new ebulds for mlir and flang is updated and PR

**Pull request closed without merging**